File: Depth-First Search/Combinations/String/LET_no17_Letter Combinations of a Phone Number.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        if not digits:
            return []
        
        memo = {"2":"abc", "3":"def", "4":"ghi", "5":"jkl", "6":"mno", "7":"pqrs", "8":"tuv", "9":"wxyz"}
        res = []
        self.helper(digits, 0, memo, [], res)
        return res

    def helper(self, s, start, memo, temp, res):
        if start == len(s):
            res.append(''.join(temp))
            return

        for c in memo[s[start]]:
            temp.append(c)
            self.helper(s, start + 1, memo, temp, res)
            temp.pop()

File: Depth-First Search/Combinations/String/LET_no17_Letter Combinations of a Phone Number.cpp
class Solution {
public:
    vector<string> letterCombinations(string digits) {
        if (digits.size() == 0)
            return {};
        
        string memo[10] = {"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
        vector<string> res;
        string temp = "";
        helper(digits, 0, memo, temp, res);
        return res;
    }

    void helper(string& digits, int start, string memo[], string& temp, vector<string>& res) {
        if (start == digits.size()) {
            res.push_back(temp);
            return;
        }

        for (char c : memo[digits[start] - '0']) {
            temp.push_back(c);
            helper(digits, start + 1, memo, temp, res);
            temp.pop_back();
        }
    }
};